database/optionally.go

```diff
@@ -0,0 +1,130 @@
+package database
+
+import (
+	"context"
+	"fmt"
+	"github.com/icinga/icinga-go-library/com"
+	"github.com/pkg/errors"
+)
+
+// Upsert upserts entities to the database.
+type Upsert interface {
+	// Stream bulk upserts the specified entities via NamedBulkExec.
+	// The upsert statement is created using BuildUpsertStmt with the
+	// first entity from the entities stream.
+	Stream(ctx context.Context, entities <-chan Entity) error
+}
+
+type ExecOption func(u *upsert)
+
+// WithOnUpsert adds callback for successful (bulk) DML operations.
+func WithOnUpsert(onUpsert ...OnSuccess[Entity]) ExecOption {
+	return func(u *upsert) {
+		u.onUpsert = onUpsert
+	}
+}
+
+// WithStatement uses own instead of the standard upsert statement
+func WithStatement(stmt string, placeholders int) ExecOption {
+	return func(u *upsert) {
+		u.stmt = stmt
+		u.placeholders = placeholders
+	}
+}
+
+// NewUpsert creates a new Upsert initalized with a database.
+func NewUpsert(db *DB, options ...ExecOption) Upsert {
+	u := &upsert{db: db}
+
+	for _, option := range options {
+		option(u)
+	}
+
+	return u
+}
+
+type upsert struct {
+	db           *DB
+	onUpsert     []OnSuccess[Entity]
+	stmt         string
+	placeholders int
+}
+
+func (u *upsert) Stream(ctx context.Context, entities <-chan Entity) error {
+	first, forward, err := com.CopyFirst(ctx, entities)
+	if err != nil {
+		return errors.Wrap(err, "can't copy first entity")
+	}
+
+	sem := u.db.GetSemaphoreForTable(TableName(first))
+	var stmt string
+	var placeholders int
+
+	if u.stmt != "" {
+		stmt = u.stmt
+		placeholders = u.placeholders
+	} else {
+		stmt, placeholders = u.db.BuildUpsertStmt(first)
+	}
+
+	return u.db.NamedBulkExec(
+		ctx, stmt, u.db.BatchSizeByPlaceholders(placeholders), sem,
+		forward, SplitOnDupId[Entity], u.onUpsert...,
+	)
+}
+
+// Delete deletes entities from the database.
+type Delete interface {
+	// Stream bulk deletes the specified args via BulkExec.
```

Comment:
Stream bulk deletes rows from the table specified in from using the given args stream via BulkExec. Unless explicitly specified, the DELETE statement is created using BuildDeleteStmt.